pkg/share/manager/jsoncs3/providercache/providercache.go

```diff
@@ -146,221 +158,336 @@ func (c *Cache) Add(ctx context.Context, storageID, spaceID, shareID string, sha
-}
-
 func (c *Cache) syncWithLock(ctx context.Context, storageID, spaceID string) error {
-	ctx, span := appctx.GetTracerProvider(ctx).Tracer(tracerName).Start(ctx, "Sync")
+	ctx, span := tracer.Start(ctx, "syncWithLock")
 	defer span.End()
 
-	span.SetAttributes(attribute.String("cs3.storageid", storageID), attribute.String("cs3.spaceid", spaceID))
+	c.initializeIfNeeded(storageID, spaceID)
 
-	log := appctx.GetLogger(ctx).With().Str("storageID", storageID).Str("spaceID", spaceID).Logger()
+	spaces, _ := c.Providers.Load(storageID)
+	space, _ := spaces.Spaces.Load(spaceID)
+	span.SetAttributes(attribute.String("cs3.storageid", storageID), attribute.String("cs3.spaceid", spaceID), attribute.String("etag", space.Etag))
+	log := appctx.GetLogger(ctx).With().Str("storageID", storageID).Str("spaceID", spaceID).Str("etag", space.Etag).Str("hostname", os.Getenv("HOSTNAME")).Logger()
 
-	var mtime time.Time
-	if c.Providers[storageID] != nil && c.Providers[storageID].Spaces[spaceID] != nil {
-		mtime = c.Providers[storageID].Spaces[spaceID].Mtime
+	dlreq := metadata.DownloadRequest{
+		Path: spaceJSONPath(storageID, spaceID),
+	}
+	// when we know an etag, only download if it changed remotely
+	if space.Etag != "" {
+		dlreq.IfNoneMatch = []string{space.Etag}
+	}
 
-		if time.Now().Before(c.Providers[storageID].Spaces[spaceID].nextSync) {
-			span.AddEvent("skip sync")
-			span.SetStatus(codes.Ok, "")
+	var dlres *metadata.DownloadResponse
+	var err error
+	downloadFunc := func() error {
+		dlres, err = c.storage.Download(ctx, dlreq)
+		switch err.(type) {
+		case nil:
+			return nil
+		case errtypes.NotFound:
+			span.AddEvent("not found")
+			return nil
+		case errtypes.NotModified:
+			span.AddEvent("not modified")
 			return nil
+		default:
+			span.RecordError(err)
+			span.SetStatus(codes.Error, "downloading provider cache failed")
+			return err
 		}
-		c.Providers[storageID].Spaces[spaceID].nextSync = time.Now().Add(c.ttl)
-	} else {
-		mtime = time.Time{} // Set zero time so that data from storage always takes precedence
 	}
-
-	jsonPath := spaceJSONPath(storageID, spaceID)
-	info, err := c.storage.Stat(ctx, jsonPath)
+	err = downloadFunc()
 	if err != nil {
-		if _, ok := err.(errtypes.NotFound); ok {
-			span.AddEvent("no file")
-			span.SetStatus(codes.Ok, "")
-			return nil // Nothing to sync against
-		}
-		if _, ok := err.(*os.PathError); ok {
-			span.AddEvent("no dir")
-			span.SetStatus(codes.Ok, "")
-			return nil // Nothing to sync against
-		}
-		span.SetStatus(codes.Error, fmt.Sprintf("Failed to stat the provider cache: %s", err.Error()))
-		log.Error().Err(err).Msg("Failed to stat the provider cache")
-		return err
-	}
-	// check mtime of /users/{userid}/created.json
-	if utils.TSToTime(info.Mtime).After(mtime) {
-		span.AddEvent("updating cache")
-		//  - update cached list of created shares for the user in memory if changed
-		createdBlob, err := c.storage.SimpleDownload(ctx, jsonPath)
+		err = downloadFunc()
```

Comment:
Shouldn't we have some retry logic here? We execute the same func without waiting twice. Why would we expect different results? Why do we retry only once?

There a multiple occurences of this, so I assume this is a deliberate decision?

Reply:
NotFound or NotModified errors are expected as part of normal operations. Only for the other errors are we going to retry the download. In theory that should not be necessary at all. We are just being cautious.

Reply:
Not talking about the error switch, thats fine and clear imo. I'm about:
```
	err = downloadFunc()
	if err != nil {
		err = downloadFunc()
       }
```
We are executing the same func without any checks twice. There is also no retry counter. So we just retry twice within milliseconds.

Reply:
cleaned up and tested locally and with the helm chart